### PR TITLE
Add gulp to build process

### PR DESCRIPTION
**Pull request by GradyD**
With the other gulp PR's having issues, merge conflicts, and/or massive changes I decided to take another stab and keep my PR very small and lite. This PR renames the source from web to web-src and creates the gulp build file. I tried to make all of the commits short and with a descriptive name of exactly what is going on, which is why there are so many. 

The biggest change is that build.properties no longer includes any of the bower_resources because gulp copies them all into the web folder. 

This PR does not include the CSS to SASS, gulp serve, or removing bower

Please let me know if there are any questions/concerns, I would like to get this merged in ASAP so that I can continue with more improvements that build on top of this PR.

### CODE REVIEW

**Comment by maggu2810** [comment]
> I would like to get this merged in ASAP so that I can continue with more improvements that build on top of this PR.

@hoegertn, @sja: all fine with this PR?

**Comment by maggu2810** [comment]
@kaikreuzer I assume we need some feedback of "the web guys" :wink:

**Comment by hoegertn** [comment]
I do not really understand what the purpose of this new PR is as it renders all the work in the other ones useless imho. Why can't we just merge the existing ones?

**Comment by maggu2810** [comment]
Don't know. I've lost track. There are now a few PRs for the rework of the Paper UI build process. As I am not familiar with all that web stuff, I hope to get some information which one is still needed and which one is the one that should be merged at all.

**Comment by hoegertn** [comment]
As I see it #828 contains all the other relevant PR. If the only problem are the merge conflicts I can have a look at it and make it mergable. BTW I have no idea why the ip checks fail as I signed the licensing stuff.

**Comment by kaikreuzer** [comment]
> I assume we need some feedback of "the web guys"

Yes, I am also leaving it to @sja and @aounhaider1 to give a "lgtm" - I am lost in all these PRs as well...

File: extensions/ui/org.eclipse.smarthome.ui.paper/gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+            // @TODO Figure out why including angular in concat breaks stuff
+            //'./web-src/bower_components/angular/angular.min.js',
+            './web-src/bower_components/angular-route/angular-route.min.js',
+            './web-src/bower_components/angular-resource/angular-resource.min.js',
+            './web-src/bower_components/angular-animate/angular-animate.min.js',
+            './web-src/bower_components/angular-aria/angular-aria.min.js',
+            './web-src/bower_components/angular-aria/angular-aria.min.js',
+            './web-src/bower_components/angular-material/angular-material.min.js',
+        ],
+        'name': 'angular-resources.js'
```
Comment (sja) [inline extensions/ui/org.eclipse.smarthome.ui.paper/gulpfile.js]: Please name this "angular-bundle.js" to not confuse with ngResource.

**Comment by GradyD** [comment]
I personally do not have a preference of which one gets merged in. The goal I had when creating this PR was to make a very 'lite' PR that can easily be merged in. The other PRs have massive changes (1,000+) with lots of changes, which can be time consuming and hard to review. I am aware I contributed to the massive changes by trying to lump all my changes together instead of working in bite sized pieces. This PR I feel, and tried, to make it very targeted, this is my goal with all future PRs so that they can, hopefully, be very quickly merged in and help move in an agile type of style. Whichever one is merged in we can easily build upon and will accomplish the same goal(s) either way in the long term.

I will say that there are two different approaches, and if there was another dev working on this there would probably a 3rd approach. I don't care which approach is taken as they both offer similar benefits and both have their own downsides, I just want progress. I will let the core team decide which approach they feel is better for the overall project and work from there. If they feel that #828 is the best approach I will gladly work off of that. I just want one of them to be merged in so that we can continue to make progress and move the entire project forward.

@sja I made the change you asked and I broke the IP check with the latest commit even though I signed off on it, sorry about that.

**Comment by kaikreuzer** [comment]
> With the other gulp PR's having issues, merge conflicts, and/or massive changes I decided to take another stab and keep my PR very small and lite

Does this mean that we can close the other PRs as they are outdated?

**Comment by hoegertn** [comment]
they add functionality that is not included in this PR but I think it is easier to fix the old PR than to redo the whole work on top of this PR. But if you decide to merge this one I can have a look at doing the changes again but this will take some time.

File: extensions/ui/org.eclipse.smarthome.ui.paper/web-src/index.html [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+		</header>
+
+		<ng-view autoscroll="true"></ng-view>
+
+		<footer> </footer>
+	</div>
+
+	<!-- lib js files -->
+	<script src="js/angular.min.js"></script>
+	<script src="js/angular-resources.min.js"></script>
```
Comment (aounhaider1) [inline extensions/ui/org.eclipse.smarthome.ui.paper/web-src/index.html]: Please update it to ```angular-bundle.js```

**Comment by aounhaider1** [comment]
Overall this pull request looks better. It incorporates the feedback given here https://github.com/eclipse/smarthome/pull/800#issuecomment-170221083 by https://github.com/sja. However, I am concerned about the front-end development after we start using Gulp. For development, I believe, we will have to switch to 'web-src' directory as minified files in 'web' cannot be debugged properly. For that, I think, its better if we keep the source and destination file names same (no *.min.js). So that we don't have to change them every time we switch between development and production mode.
Also please rebase this on the current master. The conflicts are due to https://github.com/eclipse/smarthome/pull/819.

**Comment by maggu2810** [comment]
> minified files in 'web' cannot be debugged properly

Would it be able to disable / enable the minification, so it could be controlled on build time?

**Comment by kaikreuzer** [comment]
Even then the roundtrip is annoying, since you would want to work directly on the sources during development. But maybe there are best practices for this in web development that we are not aware of?

**Comment by maggu2810** [comment]
> Even then the roundtrip is annoying, since you would want to work directly on the sources during development.

I would keep it disabled all the time except for "release" (stuff that is transmitted to a mvn / p2 repo).

> But maybe there are best practices for this in web development that we are not aware of?

:+1: 
I don't know anything about that stuff :wink:

**Comment by hoegertn** [comment]
All this is exactly what the #828 is about.

**Comment by GradyD** [comment]
@aounhaider1 Sorry about that, forgot to make that update. I will make the update this evening (I'm EST time).

As far as the using a dist vs develop I can easily accomplish this. I will update the gulp file to include these changes also. The one thing that I don't know how to accomplish is using maven to build for developing instead of dist, should this be something we look into?

@hoegertn I have locally on my computer created branches that will include the SASS and gulp serve/watch changes. This should be all of your changes except for the injection. I personally am not a fan of using injection which is why I did not spend time on it.

All that said, if you look at PR #821 with what @hoegertn has done it does accomplish everything that I planned to accomplish in the long term. The only difference is our approach and instead of having a massive PR I tried to break them apart into smaller PRs so that they are easier to review, this PR is what I would consider a base with more to follow once it is accepted. 

The biggest difference between the two is the approach and from what testing I have done on #821 it works just as well as this one.

**Comment by kaikreuzer** [comment]
> All this is exactly what the #828 is about.

Ok, even without #828 in place, this PR is fine and development could be done by serving the web-src instead of the web folder, if the ".min" is removed from the js files, right? I think this is the same suggestion as from @sja in https://github.com/eclipse/smarthome/pull/800#issuecomment-170221083.

So if I understand it right, once you have done the last changes, this PR is fine for everybody and can be merged.

**Comment by GradyD** [comment]
I've fixed the merge conflict as well as fixing the issue @aounhaider1 caught.

At this point I do not have the ability to do the mixed mode resources. I spent a good bit of time reading blogs and best practices on how to accomplish this. 

One option is to use what @hoegertn has done using the gulp injection plugin. It appears easy enough to add in but doing so will require quite a few changes to the gulp file so I would like to push that to a second PR. Since it will be the only change made in that PR it should be very easy to review and accept. The one downside to using injection is that your final markup of `index.html` is now 100% dependent upon the build process. This is the biggest reason I am not a fan of it.

We can do what @kaikreuzer mentioned of using the same file name between web-src and web the downside to this is that we are still combining multiple files together into 1 file so you will have to dig through to find which file you need to edit.

The other option is to use sourcemaps, [blog.teamtreehouse](http://blog.teamtreehouse.com/introduction-source-maps). Assuming I am reading and understanding sourcemaps correctly it would be as simple as having a flag when we minify the JS files. Nothing extra would need to be added to the `index.html` as the reference to the sourcemap is at the end of the JS file and is only loaded if you open your dev console and have sourcemaps turned on in your browser. What the sourcemap will do is show the unminified code in your console allowing you to debug and view the code as if it was never minified, even though it is still minified. The downside is I did see where some maps got huge and it would take a few seconds to load in and we would need to make sure they are not packaged into the the file jar just to not add bloat.

My vote is to go with the sourcemaps. @sja, @aounhaider1, @kaikreuzer, @hoegertn what are your thoughts?

Does the 'develop' mode need to be added to this PR or can it wait and be added with a future PR? If this can wait for a different PR I will make sure it is at the top of my TODO list as well as gulp serve so that we can start making changes without having to deal with building the jar every time. If we feel it needs to be added here lets finalize on the route we feel is best and I will work to get it added ASAP.

**Comment by kaikreuzer** [comment]
> Does the 'develop' mode need to be added to this PR or can it wait and be added with a future PR? 

As @aounhaider1 is the only one heavily working on implementing new features in the Paper UI and thus is impacted by this, I leave that decision to him. A possible workaround for the time being might be to simply maintain a second index.html for development purposes.

**Comment by aounhaider1** [comment]
I think for now we can move forward with this pull request. As a next step we should try to improve the gulp build process. @GradyD I will also look into gulp injection plugin and sourcemaps, so that we can make a more informed decision on what should be done next.

**Comment by kaikreuzer** [comment]
Alright then - @GradyD would you please squash your commits into one, then I will merge the PR!

**Comment by GradyD** [comment]
Ok. I have squashed all my commits.

Lets merge :+1:

**Pull request merged**